File: v1/examples/csharp/10_Parking.cs
// 10_Parking.cs - v1 API Parking Management Example
//
// Query parking lot and parking spot information and status.
//
// Endpoints:
//   GET /api/v1/parking/lot          - Parking lot list
//   GET /api/v1/parking/lot/status   - Lot status (occupancy)
//   GET /api/v1/parking/spot         - Parking spot list
//   GET /api/v1/parking/spot/status  - Spot status
//
// Note: v0 API does not have parking-related endpoints.

using System;
using System.Text.Json;
using System.Threading.Tasks;

namespace TsApiExamples.V1
{
    class Parking
    {
        static async Task Main(string[] args)
        {
            using var client = new NvrClient();
            if (string.IsNullOrEmpty(client.ApiKey))
            {
                Console.WriteLine("NVR_API_KEY environment variable is required");
                return;
            }
            client.SetApiKey(client.ApiKey);

            var options = new JsonSerializerOptions { WriteIndented = true };

            // ── Parking lot list ──
            Console.WriteLine("=== Parking Lot List ===");
            var lr = await client.GetAsync("/api/v1/parking/lot");
            if (lr.IsSuccessStatusCode)
            {
                using var ldoc = await NvrClient.ReadJsonAsync(lr);
                foreach (var lot in ldoc.RootElement.EnumerateArray())
                {
                    Console.WriteLine($"  {JsonSerializer.Serialize(lot, options)}");
                }
            }
            else
            {
                Console.WriteLine($"  Response: {(int)lr.StatusCode} {lr.StatusCode}");
                Console.WriteLine("  Parking feature may not be configured.");
            }

            // ── Parking lot status (occupancy) ──
            Console.WriteLine("\n=== Parking Lot Status ===");
            var lsr = await client.GetAsync("/api/v1/parking/lot/status");
            if (lsr.IsSuccessStatusCode)
            {
                using var lsdoc = await NvrClient.ReadJsonAsync(lsr);
                foreach (var status in lsdoc.RootElement.EnumerateArray())
                {
                    Console.WriteLine($"  {JsonSerializer.Serialize(status, options)}");
                }
            }
            else
            {
                Console.WriteLine($"  Response: {(int)lsr.StatusCode}");
            }

            // ── Parking spot list ──
            Console.WriteLine("\n=== Parking Spot List ===");
            var spr = await client.GetAsync("/api/v1/parking/spot");
            if (spr.IsSuccessStatusCode)
            {
                using var spdoc = await NvrClient.ReadJsonAsync(spr);
                var spots = spdoc.RootElement;
                Console.WriteLine($"  Total spots: {spots.GetArrayLength()}");

                // Show up to 10
                int shown = 0;
                foreach (var spot in spots.EnumerateArray())
                {
                    if (++shown > 10) break;
                    Console.WriteLine($"  [{shown}] {JsonSerializer.Serialize(spot)}");
                }
                if (spots.GetArrayLength() > 10)
                    Console.WriteLine($"  ... {spots.GetArrayLength() - 10} more");
            }
            else
            {
                Console.WriteLine($"  Response: {(int)spr.StatusCode}");
            }

            // ── Parking spot status ──
            Console.WriteLine("\n=== Parking Spot Status ===");
            var ssr = await client.GetAsync("/api/v1/parking/spot/status");
            if (ssr.IsSuccessStatusCode)
            {
                using var ssdoc = await NvrClient.ReadJsonAsync(ssr);
                var spotStatuses = ssdoc.RootElement;
                Console.WriteLine($"  Total: {spotStatuses.GetArrayLength()}");

                // Count occupied/vacant
                int occupied = 0, vacant = 0, unknown = 0;
                foreach (var s in spotStatuses.EnumerateArray())
                {
                    if (s.TryGetProperty("occupied", out var occ))
                    {
                        if (occ.GetBoolean()) occupied++;
                        else vacant++;
                    }
                    else
                    {
                        unknown++;
                    }
                }
                Console.WriteLine($"  Occupied: {occupied}");
                Console.WriteLine($"  Vacant: {vacant}");
                if (unknown > 0)
                    Console.WriteLine($"  Unknown: {unknown}");
            }
            else
            {
                Console.WriteLine($"  Response: {(int)ssr.StatusCode}");
            }
        }
    }
}

File: v1/examples/python/10_parking.py
"""
10_parking.py - TS-API v1 Parking

Endpoints:
  GET /api/v1/parking/lot           - List parking lots
  GET /api/v1/parking/lot/status    - Parking lot occupancy status
  GET /api/v1/parking/spot          - Recognition zones (all types: spot, entrance, exit, noParking, recognition)
  GET /api/v1/parking/spot/status   - Status of each parking spot
"""

import os
import sys
import json

sys.path.insert(0, os.path.dirname(os.path.dirname(os.path.abspath(__file__))))
from config import NVR_API_KEY
from http_client import NvrClient

if not NVR_API_KEY:
    print('NVR_API_KEY environment variable is required')
    sys.exit(1)


def main():
    client = NvrClient()
    client.set_api_key(NVR_API_KEY)

    # --- Parking Lots ---
    print("=== Parking Lots ===")
    r = client.get('/api/v1/parking/lot')
    if r.status_code == 200:
        lots = r.json()
        print(f"  {len(lots)} lots found")
        for lot in lots:
            info = f"  [{lot['id']}] {lot['name']} (type={lot.get('type')}, max={lot.get('maxCount')})"
            if 'parkingSpots' in lot:
                info += f" spots={lot['parkingSpots']}"
            if 'member' in lot:
                info += f" member={lot['member']}"
            print(info)
    else:
        print(f"  Status: {r.status_code}")

    # --- Parking Lot Status (occupancy) ---
    print("\n=== Parking Lot Status ===")
    r = client.get('/api/v1/parking/lot/status')
    if r.status_code == 200:
        statuses = r.json()
        for s in statuses:
            print(f"  {json.dumps(s, ensure_ascii=False)}")
    else:
        print(f"  Status: {r.status_code}")

    # --- Recognition Zones (all types: spot, entrance, exit, noParking, recognition) ---
    print("\n=== Recognition Zones ===")
    r = client.get('/api/v1/parking/spot')
    if r.status_code == 200:
        zones = r.json()
        types = {}
        for z in zones:
            t = z.get('type', 'unknown')
            types[t] = types.get(t, 0) + 1
        print(f"  {len(zones)} zones found ({', '.join(f'{k}: {v}' for k, v in types.items())})")
        for zone in zones[:10]:  # Show first 10
            print(f"  {json.dumps(zone, ensure_ascii=False)}")
        if len(zones) > 10:
            print(f"  ... and {len(zones) - 10} more zones")
    else:
        print(f"  Status: {r.status_code}")

    # --- Parking Spot Status ---
    print("\n=== Parking Spot Status ===")
    r = client.get('/api/v1/parking/spot/status')
    if r.status_code == 200:
        statuses = r.json()
        for s in statuses[:10]:  # Show first 10
            print(f"  {json.dumps(s, ensure_ascii=False)}")
        if len(statuses) > 10:
            print(f"  ... and {len(statuses) - 10} more spots")
    else:
        print(f"  Status: {r.status_code}")


if __name__ == '__main__':
    main()
